Homework 3: Implement a decision tree using the banking data provided.

Author: Jose Solomon

In [120]:
# import pandas
import pandas as pd

# Read-in sub-sample bank data...
# Note: data is separated by semicolons, not commas, so specify when reading data
bankDF = pd.read_csv('bank.csv',sep=';')

In [121]:
# 1. Use the file **bank.csv** to explore the dataset. Observe the features: Are they numbers? 
# Are they strings? Are they binary? Are they continuous?

# Look at the header of the data
bankDF.head(10)

age            job  marital  education default  balance housing loan  \
0   30     unemployed  married    primary      no     1787      no   no   
1   33       services  married  secondary      no     4789     yes  yes   
2   35     management   single   tertiary      no     1350     yes   no   
3   30     management  married   tertiary      no     1476     yes  yes   
4   59    blue-collar  married  secondary      no        0     yes   no   
5   35     management   single   tertiary      no      747      no   no   
6   36  self-employed  married   tertiary      no      307     yes   no   
7   39     technician  married  secondary      no      147     yes   no   
8   41   entrepreneur  married   tertiary      no      221     yes   no   
9   43       services  married    primary      no      -88     yes  yes   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no  
5  cellular   23   feb       141         2    176         3  failure  no  
6  cellular   14   may       341         1    330         2    other  no  
7  cellular    6   may       151         2     -1         0  unknown  no  
8   unknown   14   may        57         2     -1         0  unknown  no  
9  cellular   17   apr       313         1    147         2  failure  no

In [73]:
# Let's get some stat descriptions
bankDF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4521 entries, 0 to 4520
Data columns (total 17 columns):
age          4521 non-null int64
job          4521 non-null object
marital      4521 non-null object
education    4521 non-null object
default      4521 non-null object
balance      4521 non-null int64
housing      4521 non-null object
loan         4521 non-null object
contact      4521 non-null object
day          4521 non-null int64
month        4521 non-null object
duration     4521 non-null int64
campaign     4521 non-null int64
pdays        4521 non-null int64
previous     4521 non-null int64
poutcome     4521 non-null object
y            4521 non-null object
dtypes: int64(7), object(10)
memory usage: 635.8+ KB


So we have a grand total of 4,521 entries with 17 columns. 7 columns are numbers, (int64), and the remaining 10 are objects, most likely strings. 

From the columns, I think columns 'defualt, 'housing','loan', and 'y' are boolean, (yes/no). 'balance' can be continuous in nature, although it appersvalues have been rounded off to the nearest dollar. All other variables are descrete or string descriptors.

In [118]:
# 2. Learn about label encoders at the following link and use what you learn to transform the features
# to numerical features. 
import numpy as np

# Create an encoded data frame
encodedDF = pd.DataFrame.copy(bankDF)

# Select the columns that need to be encoded
columns = np.array([1, 2, 3, 4, 6, 7, 8, 10, 15, 16])

# I used label encoder instead of OneHotEncoder... 
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

# Go throu
for col in bankDF:
    if bankDF[col].dtype == 'O':
        le.fit(bankDF[col])
        encodedDF[col] = le.transform(bankDF[col])
        
encodedDF.head(10)
        
# # Process the dataFrame
# for i in range(0,1):
#     le.fit(bankDF.iloc[columns[i]])
#     encodedDF.iloc[columns[i]] = le.transform(bankDF.iloc[columns[i]])

# # Show resulting encoded dataframe
# encodedDF.head(30)

# # This works
# le.fit(bankDF.iloc[1])
# le.classes_
# testDF = pd.DataFrame(bankDF)
# testDF.iloc[1] = le.transform(bankDF.iloc[1])
# testDF.head(30)

# le.fit(bankDF.iloc[:,2])
# le.classes_
# testDF = pd.DataFrame(bankDF)
# testDF.iloc[:,2] = le.transform(bankDF.iloc[:,2])
# testDF.head(30)


# from sklearn.preprocessing import OneHotEncoder
# encoder = OneHotEncoder(n_values='auto', categorical_features='all', sparse=True)
# encoder.fit(bankDF)
# encoded = pd.get_dummies(bankDF)

age  job  marital  education  default  balance  housing  loan  contact  \
0   30   10        1          0        0     1787        0     0        0   
1   33    7        1          1        0     4789        1     1        0   
2   35    4        2          2        0     1350        1     0        0   
3   30    4        1          2        0     1476        1     1        2   
4   59    1        1          1        0        0        1     0        2   
5   35    4        2          2        0      747        0     0        0   
6   36    6        1          2        0      307        1     0        0   
7   39    9        1          1        0      147        1     0        0   
8   41    2        1          2        0      221        1     0        2   
9   43    7        1          0        0      -88        1     1        0   

   day  month  duration  campaign  pdays  previous  poutcome  y  
0   19     10        79         1     -1         0         3  0  
1   11      8       220         1    339         4         0  0  
2   16      0       185         1    330         1         0  0  
3    3      6       199         4     -1         0         3  0  
4    5      8       226         1     -1         0         3  0  
5   23      3       141         2    176         3         0  0  
6   14      8       341         1    330         2         1  0  
7    6      8       151         2     -1         0         3  0  
8   14      8        57         2     -1         0         3  0  
9   17      0       313         1    147         2         0  0